### PR TITLE
feat(npm): handle github dependencies with semver versions

### PR DESCRIPTION
**Pull request by fehnomenal**
## Changes

The npm manager supports extracting the correct version from strings like `github:renovatebot/renovate#semver:37.280.0`.

## Context

NPM supports those dependencies but renovate not yet.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/npm/extract/common/dependency.ts [hunk 1]
```diff
@@ -180,6 +180,16 @@ export function extractDependency(
     dep.datasource = GithubTagsDatasource.id;
     dep.versioning = npmVersioningId;
     dep.packageName = githubOwnerRepo;
+  } else if (
+    depRefPart.startsWith('semver:') &&
+    isVersion(depRefPart.slice('semver:'.length))
+  ) {
```
Comment (rarkins) [inline lib/modules/manager/npm/extract/common/dependency.ts:186]:
```suggestion
  ) {
    // <protocol>://[<user>[:<password>]@]<hostname>[:<port>][:][/]<path>[#<commit-ish> | #semver:<semver>]
    // https://docs.npmjs.com/cli/v10/configuring-npm/package-json#git-urls-as-dependencies
```

**Review by rarkins — CHANGES_REQUESTED** [review]
Also needs a test added

File: lib/modules/manager/npm/extract/common/dependency.ts [hunk 2]
```diff
@@ -180,6 +180,16 @@ export function extractDependency(
     dep.datasource = GithubTagsDatasource.id;
     dep.versioning = npmVersioningId;
     dep.packageName = githubOwnerRepo;
+  } else if (
+    depRefPart.startsWith('semver:') &&
+    isVersion(depRefPart.slice('semver:'.length))
```
Comment (viceice) [inline lib/modules/manager/npm/extract/common/dependency.ts:185]: use substring (perf)

File: lib/modules/manager/npm/extract/common/dependency.ts [hunk 3]
```diff
@@ -180,6 +180,16 @@ export function extractDependency(
...
+  ) {
+    dep.currentRawValue = dep.currentValue;
+    dep.currentValue = depRefPart.slice('semver:'.length);
```
Comment (viceice) [inline lib/modules/manager/npm/extract/common/dependency.ts:188]: same

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs tests added

File: lib/modules/manager/npm/extract/common/dependency.ts (hunk 1 above)
Comment (fehnomenal) [reply lib/modules/manager/npm/extract/common/dependency.ts]: Done.

File: lib/modules/manager/npm/extract/common/dependency.ts (hunk 2 above)
Comment (fehnomenal) [reply lib/modules/manager/npm/extract/common/dependency.ts]: I was not aware of a performance difference but I updated the code to only extract the version once.

**Comment by fehnomenal** [comment]
I merged main and added tests. Thank you for reviewing this PR

File: lib/modules/manager/npm/extract/common/dependency.ts [hunk 4]
```diff
@@ -181,8 +181,22 @@ export function extractDependency(
...
   } else {
-    dep.skipReason = 'unversioned-reference';
-    return dep;
+    // <protocol>://[<user>[:<password>]@]<hostname>[:<port>][:][/]<path>[#<commit-ish> | #semver:<semver>]
+    // https://docs.npmjs.com/cli/v10/configuring-npm/package-json#git-urls-as-dependencies
+    const len = 7; // length of 'semver:'
+    const maybeVersion = depRefPart.substring(len);
+
+    if (depRefPart.startsWith('semver:') && isValid(maybeVersion)) {
+      dep.currentRawValue = dep.currentValue;
```
Comment (viceice) [inline lib/modules/manager/npm/extract/common/dependency.ts:190]: what is `currenRawValue` ? 🤔

I've never seen it before.

@rarkins do you know something? or is it totally wrong?

Comment (fehnomenal) [reply lib/modules/manager/npm/extract/common/dependency.ts:190]: Well, it is assigned in all other cases as well. That's why I copied it.

Comment (rarkins) [reply lib/modules/manager/npm/extract/common/dependency.ts:190]: Yeah, let's keep it - it's used

**Review by rarkins — APPROVED** [review]

**Pull request merged**